content/en/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

```diff
@@ -0,0 +1,300 @@
+
+## Typical use cases
+
+Below are some real-life scenarios where control over the scheduling and eviction of pods may be required.
+
+1. Let's say the pod you are planning to deploy is really critical, and you have some resource constraints. How will you make the resources available in the cluster? by first identifying which pods are not required, then manually deleting them, etc., which is a manual and tedious process, or you may plan to add a new node to the cluster, which too might take some time and money. 
+
+   An example of such a case is a DaemonSet of an infrastructure component like Grafana Loki. In that case, the Loki pods must run on each node before any other pods run on that node.
+
+2. Another use case can be: Suppose, you have a single cluster, and that cluster holds the pods for the below environments:
+   - preprod [Priority 1] 
+   - staging [Priority 2] 
+   - development [Priority 3]
+
+   Unfortunately, the cluster goes through high resource consumption, and now there is a resource competition on the nodes with respect to the CPU and the memory. At least in the short term, you don’t have further nodes to expand or scale that Kubernetes cluster (cluster level autoscaling _may_ add more nodes, but this takes time). Perhaps some pods will remain in a Pending state, or perhaps service will degrade as Pods compete for resources. If the kubelet does evict a Pod from a node, that eviction happens at random because the kubelet doesn’t have any special information about which Pods to evict and which to keep.
+
+3. Assume you have a microservices architecture that includes Kafka, Zookeeper, MongoDB, a couple of jobs, a couple of CronJobs, and around 40+ microservices. Assume that if we are caught up in this situation of resource crunch and somehow Kafka or MongoDB gets evicted, then all our microservices will be useless until and unless these pods are up and start serving traffic again. 
+
+There can also be other scenarios where you want to control the order of scheduling or order of eviction of pods.
+
+## PriorityClasses in Kubernetes
+
+PriorityClass is a cluster-wide API object in Kubernetes and part of the `scheduling.k8s.io/v1` API group. It contains a mapping of the PriorityClass name (defined in `.metadata.name`) and an integer value (defined in `.value`). This represents the value that the scheduler uses to determine Pod's relative priority.
+
+Just to add when you create a cluster using kubeadm, or even use a managed Kubernetes service (for example, Azure Kubernetes Service), Kubernetes uses PriorityClasses to safeguard the pods that are hosted on the controlplane nodes. This ensures that critical cluster components such as CoreDNS and kube-proxy are able to run even if resources are constrained.
+
+It is achieved by using a special PriorityClass that ensures that these pods can serve the requests and that the overall cluster is up and running.
```

Comment:
What is achieved? And by who/what?

Reply:
Updated: 

This availability of pods is achieved through the use of a special PriorityClass that ensures the pods are up and running and that the overall cluster is not affected.